### PR TITLE
Dylan Finlay Bootcamp Submission

### PR DESCRIPTION
**Pull request by DylanFinlay**
Please let me know if any changes are required.

### CODE REVIEW

File: modules/bootcamp/detect_landing_pad.py [hunk 1]
```diff
@@ -84,32 +84,39 @@ def run(self, image: np.ndarray) -> "tuple[list[bounding_box.BoundingBox], np.nd
...
+        predictions = self.__model.predict(source = image, conf = 0.7, device = self.__DEVICE, verbose=False)
 
         # Get the Result object
-        prediction = ...
+        prediction = predictions[0]
 
         # Plot the annotated image from the Result object
         # Include the confidence value
-        image_annotated = ...
+        image_annotated = prediction.plot()
```
Comment (Xierumeng) [inline modules/bootcamp/detect_landing_pad.py:94]: Do the logged images have the confidence labels?

File: modules/bootcamp/detect_landing_pad.py [hunk 2]
```diff
@@ -84,32 +84,39 @@ def run(self, image: np.ndarray) -> "tuple[list[bounding_box.BoundingBox], np.nd
...
         # move the copy into CPU space,
         # and convert to a numpy array
-        boxes_cpu = ...
+        boxes_cpu = boxes_xyxy.detach().cpu().numpy()
 
         # Loop over the boxes list and create a list of bounding boxes
         bounding_boxes = []
+        
+        for i in range(boxes_cpu.shape[0]):
+            box=bounding_box.BoundingBox.create(boxes_cpu[i])[1]
```
Comment (Xierumeng) [inline modules/bootcamp/detect_landing_pad.py:108]: What if box is `None` ?

File: modules/bootcamp/decision_simple_waypoint.py [hunk 3]
```diff
@@ -68,10 +71,20 @@ def run(self,
         # ↓ BOOTCAMPERS MODIFY BELOW THIS COMMENT ↓
         # ============
 
+        # Check if the drone has left the initial position and if it has not, send a command to go to the waypoint
+        if report.status == drone_status.DroneStatus.HALTED and self.has_left == False:
+            command = commands.Command.create_set_relative_destination_command(self.waypoint.location_x, self.waypoint.location_y)
+            self.has_left = True
+
+        # Check if the drone has reached the waypoint and if it has, send a command to land
+        elif report.status == drone_status.DroneStatus.HALTED and self.has_left == True:
```
Comment (Xierumeng) [inline modules/bootcamp/decision_simple_waypoint.py:80]: What if something causes the drone to halt while it's on the way to the waypoint? (You do not need to change anything here, just something to think about).

File: modules/bootcamp/decision_simple_waypoint.py [hunk 4]
```diff
@@ -68,10 +71,20 @@ def run(self,
         # ↓ BOOTCAMPERS MODIFY BELOW THIS COMMENT ↓
         # ============
 
+        # Check if the drone has left the initial position and if it has not, send a command to go to the waypoint
+        if report.status == drone_status.DroneStatus.HALTED and self.has_left == False:
```
Comment (Xierumeng) [inline modules/bootcamp/decision_simple_waypoint.py:75]: Prefer `not self.has_left` (C++ equivalent: `!self.has_left` ).

File: modules/bootcamp/decision_simple_waypoint.py (hunk 3 above)
Comment (Xierumeng) [inline modules/bootcamp/decision_simple_waypoint.py:80]: Prefer just `self.has_left` over `self.has_left == True` .

File: modules/bootcamp/decision_waypoint_landing_pads.py [hunk 5]
```diff
@@ -70,8 +73,47 @@ def run(self,
...
         # Do something based on the report and the state of this class...
 
+        # Helper function to calculate distance between two points
+        def distance_between_points(current_location: location.Location, destination_location: location.Location) -> float:
+            x1 = current_location.location_x
+            y1 = current_location.location_y
+            x2 = destination_location.location_x
+            y2 = destination_location.location_y
+
+            return ((x1 - x2)**2 + (y1 - y2)**2)**0.5
```
Comment (Xierumeng) [inline modules/bootcamp/decision_waypoint_landing_pads.py:83]: Square roots are computationally expensive, would it be possible to avoid using it? (You do not need to change anything here, just something to think about).

File: modules/bootcamp/decision_simple_waypoint.py [hunk 6]
```diff
@@ -68,10 +71,20 @@ def run(self,
...
+            command = commands.Command.create_set_relative_destination_command(self.waypoint.location_x, self.waypoint.location_y)
```
Comment (Xierumeng) [inline modules/bootcamp/decision_simple_waypoint.py:76]: What if the drone's initial position wasn't (0,0) ?

File: modules/bootcamp/decision_waypoint_landing_pads.py (hunk 5 above)
Comment (Xierumeng) [inline modules/bootcamp/decision_waypoint_landing_pads.py:83]: Square roots are computationally expensive, would it be possible to avoid using it? (You do not need to change anything here, just something to think about).

File: modules/bootcamp/decision_waypoint_landing_pads.py [hunk 7]
```diff
@@ -70,8 +73,47 @@ def run(self,
...
+        if report.status == drone_status.DroneStatus.HALTED and self.has_left_home == False and self.has_left_waypoint == False:
+            command = commands.Command.create_set_relative_destination_command(self.waypoint.location_x, self.waypoint.location_y)
+            self.has_left_home = True
+
+        # If the drone is at the waypoint, move to the nearest landing pad
+        elif report.status == drone_status.DroneStatus.HALTED and self.has_left_home == True and self.has_left_waypoint == False:
+            closest_landing_pad = landing_pad_locations[0]
+
+            for landing_pad in landing_pad_locations:
+                if distance_between_points(self.waypoint, landing_pad) < distance_between_points(self.waypoint, closest_landing_pad):
```
Comment (Xierumeng) [inline modules/bootcamp/decision_waypoint_landing_pads.py:104]: Is it necessary to calculate the distance to the closest landing pad every time? Seems rather inefficient.

Comment (Xierumeng) [inline modules/bootcamp/decision_waypoint_landing_pads.py:104]: If there's only 1 landing pad, is it necessary to recalculate the closest landing pad?

Comment (Xierumeng) [inline modules/bootcamp/decision_waypoint_landing_pads.py:104]: Good that you chose to use the waypoint instead of the drone's current position!

File: modules/bootcamp/decision_waypoint_landing_pads.py [hunk 8]
```diff
@@ -70,8 +73,47 @@ def run(self,
...
+                    closest_landing_pad = landing_pad
+            
+            command = commands.Command.create_set_relative_destination_command(relative_destination(report.position, closest_landing_pad).location_x, relative_destination(report.position, closest_landing_pad).location_y)
```
Comment (Xierumeng) [inline modules/bootcamp/decision_waypoint_landing_pads.py:107]: This is hard to read (and inefficient). Better to assign the result of `relative_destination()` to a variable to avoid calling it twice (and avoid creating 2 objects instead of 1).

File: modules/bootcamp/decision_waypoint_landing_pads.py [hunk 9]
```diff
@@ -70,8 +73,50 @@ def run(self,
...
+        def relative_destination(current_location: location.Location, destination_location: location.Location) -> location.Location:
+            x1 = current_location.location_x
+            y1 = current_location.location_y
+            x2 = destination_location.location_x
+            y2 = destination_location.location_y
+
+            return location.Location(x2 - x1, y2 - y1)
+
+        # If the drone is not at the waypoint, move to the waypoint
+        if report.status == drone_status.DroneStatus.HALTED and self.has_left_home == False and self.has_left_waypoint == False:
```
Comment (Xierumeng) [inline modules/bootcamp/decision_waypoint_landing_pads.py:95]: `not` preferred.

File: modules/bootcamp/decision_waypoint_landing_pads.py [hunk 10]
```diff
@@ -70,8 +73,50 @@ def run(self,
...
+            command = commands.Command.create_set_relative_destination_command(self.waypoint.location_x - report.position.location_x, self.waypoint.location_y - report.position.location_y)
+            self.has_left_home = True
+
+        # If the drone is at the waypoint, move to the nearest landing pad
+        elif report.status == drone_status.DroneStatus.HALTED and self.has_left_home == True and self.has_left_waypoint == False:
```
Comment (Xierumeng) [inline modules/bootcamp/decision_waypoint_landing_pads.py:100]: `not` preferred.

File: modules/bootcamp/decision_waypoint_landing_pads.py [hunk 11]
```diff
@@ -70,8 +73,50 @@ def run(self,
...
+            closest_landing_pad = landing_pad_locations[0]
```
Comment (Xierumeng) [inline modules/bootcamp/decision_waypoint_landing_pads.py:101]: While the bootcamp specifies at least 1 landing pad, what happens if there were no landing pads? (You do not need to change anything here, just something to think about).

File: modules/bootcamp/decision_waypoint_landing_pads.py [hunk 12]
```diff
@@ -70,8 +73,50 @@ def run(self,
...
+            closest_pad_dist = distance_between_points(self.waypoint, closest_landing_pad)
+
+            # Check if the below works
+            for landing_pad in landing_pad_locations[1:]:
```
Comment (Xierumeng) [inline modules/bootcamp/decision_waypoint_landing_pads.py:105]: Slicing, nice.

**Review by Xierumeng — COMMENTED** [review]
LGTM.
